AVM: Implement lsig size pooling

giuliop opened the pull request:
# Implement pooling of logic signatures' size across a transaction group.
This PR addresses this [issue](https://github.com/algorand/go-algorand/issues/5990) by implementing pooling of logic signatures' size across a transaction group.  
As discussed in the [issue](https://github.com/algorand/go-algorand/issues/5990), given that currently it is possible to submit a group with 16 1KB lsigs, it seems logical to allow pooling of the total size across a group.

### Implementation
The pooled max lsig length for the group is set at `len(group) * LogicSigMaxSize`, as done for lsig [opcode budget pooling](https://github.com/algorand/go-algorand/blob/master/data/transactions/logic/eval.go#L434), for both [node validation](https://github.com/algorand/go-algorand/compare/master...giuliop:pool-max-logicSig-size?expand=1#diff-d3741bc12fa9faf0a148f027517dbfde4942668add6ee4ac7b4032543db4b0f5R232) and [goal clerk dryrun](https://github.com/algorand/go-algorand/compare/master...giuliop:pool-max-logicSig-size?expand=1#diff-93be0b988ece327bf5a7be6c6439bcc5d1d8b40edd7250986883f4fe779e41b6R1212).

A group is valid if the sum of all lsigs' program and args size in the group does not exceed the pooled max lsig length.

The [consensus change](https://github.com/giuliop/go-algorand/commit/9d7d6a64488a5caf5fda01f0a82c3eb488fe92e0#diff-a75ef7294adbd7ae646c49ae2a518253936983c0a16594b84f20f15403767a2fR121)(`EnableLogicSigSizePooling bool`) is enabled in [vFuture](https://github.com/giuliop/go-algorand/commit/9d7d6a64488a5caf5fda01f0a82c3eb488fe92e0#diff-a75ef7294adbd7ae646c49ae2a518253936983c0a16594b84f20f15403767a2fR1519) in this PR.

To make pass the [TestMaxSizesCorrect](https://github.com/algorand/go-algorand/blob/b3c7bca66d3dcbccfd5cf7abc5d30620b175406b/node/node_test.go#L783), I had to change the value of [TxnTagMaxSize](https://github.com/algorand/go-algorand/compare/master...giuliop:pool-max-logicSig-size?expand=1#diff-9a5f16d58145af003ece0a88e66e2db2e9014466ac167ff47265bc9a0e595e19R90).  
Two observations on this:
1) The comments of TestMaxSizesCorrect say:
```
// If this test fails, DO NOT JUST UPDATE THE CONSTANTS OR MODIFY THE TEST!
// Instead you need to introduce a new version of the protocol and mechanisms
// to ensure that nodes on different proto versions don't reject each others messages due to exceeding
// max size network protocol version
```
Beside updating the constant I did introduce a new version of the protocol (in vFuture) but I don't fully understand the comment so not sure if I was supposed to do something else as well. 

2) To calculate the value of [TxnTagMaxSize](https://github.com/algorand/go-algorand/blob/b3c7bca66d3dcbccfd5cf7abc5d30620b175406b/node/node_test.go#L783), [TestMaxSizesCorrect](https://github.com/algorand/go-algorand/blob/b3c7bca66d3dcbccfd5cf7abc5d30620b175406b/node/node_test.go#L783) calls the function [SignedTxnMaxSize](https://github.com/algorand/go-algorand/blob/b3c7bca66d3dcbccfd5cf7abc5d30620b175406b/node/node_test.go#L808) which calculates the `LogicSigMaxSize` calling this function:
<pre>
// MaxSize returns a maximum valid message size for this message type
func LogicSigMaxSize() (s int) {
	s = 1 + 2 + msgp.BytesPrefixSize + config.MaxLogicSigMaxSize + 4 + crypto.SignatureMaxSize() + 5 + crypto.MultisigSigMaxSize() + 4
	// Calculating size of slice: z.Args
	s += msgp.ArrayHeaderSize + ((<b>EvalMaxArgs</b>) * (msgp.BytesPrefixSize + <b>config.MaxLogicSigMaxSize</b>))
	return
}
</pre>
`config.MaxLogicSigMaxSize` used to be 1KB and this PR changes it to [16KB](https://github.com/algorand/go-algorand/compare/master...giuliop:pool-max-logicSig-size?expand=1#diff-a75ef7294adbd7ae646c49ae2a518253936983c0a16594b84f20f15403767a2fR772).  
What I don't undestand is why we multiply `config.MaxLogicSigMaxSize` by `EvalMaxArgs` which is [255](https://github.com/algorand/go-algorand/blob/b3c7bca66d3dcbccfd5cf7abc5d30620b175406b/data/transactions/logicsig.go#L26), ending up with what looks like an exagerrated value for `LogicSigMaxSize`.  
I did not want to touch this in case I am missing something.

### Testing plan
The PR is marked as draft since it does not includes tests yet.  
I've run separately [integration tests](https://github.com/giuliop/test-lsig-size-pooling) testing success and failure for single transactions and groups, both for `goal clerk dryrun` and actually sending the transactions.

Before integrating the tests in the codebase I'd like to get some feedback:
1) Is the overall approach ok?
2) How would you like to see the tests? I suppose integration tests only? Any guidance on where to put the tests and which existing tests to use as reference would be very helpful

giuliop committed:
Implement lsig size pooling

giuliop committed:
Merge remote-tracking branch 'upstream/master' into pool-max-logicSig-size

jannotti commented on data/transactions/verify/txn.go:
```diff
@@ -225,6 +226,18 @@ func txnGroupBatchPrep(stxs []transactions.SignedTxn, contextHdr *bookkeeping.Bl
+				"txgroup had %d bytes of LogicSigs, more than the available pool of %d bytes",
+				lSigPooledSize, lSigMaxPooledSize,
+			)
+			err = &TxGroupError{err: errors.New(errorMsg), GroupIndex: -1, Reason: TxGroupErrorReasonLogicSigFailed}
```
We need a different `Reason` here, since this counter seems designed to count how often LogicSigs _execute_ but fail.  I suppose maybe `TxGroupErrorReasonNotWellFormed` would be ok, rather than introduce a new one.

giuliop replied:
I agree, I also changed this [comment](https://github.com/algorand/go-algorand/pull/6057/files#diff-d3741bc12fa9faf0a148f027517dbfde4942668add6ee4ac7b4032543db4b0f5R96) for consistency, since `txn.WellFormed` does not fail on this

jannotti commented on data/transactions/verify/txn.go:
```diff
@@ -361,7 +374,7 @@ func logicSigSanityCheckBatchPrep(gi int, groupCtx *GroupContext, batchVerifier
 		return errors.New("LogicSig.Logic version too new")
 	}
-	if uint64(lsig.Len()) > groupCtx.consensusParams.LogicSigMaxSize {
+	if !groupCtx.consensusParams.EnableLogicSigSizePooling && uint64(lsig.Len()) > groupCtx.consensusParams.LogicSigMaxSize {
```
Since you're here, I'd prefer this error message change to `"LogicSig too long"` since `Len()` is (correctly) reporting the sum of `Logic` and `Args`

giuliop replied:
you are correct, changed

jannotti commented:
I tentatively agree with you about you thinking about `LogicSigMaxSize()`. There is no explicit constant for a maximum argument size, so an argument can indeed by the whole 16kb in size. But, of course, you can't have 255 of them.  So I think we want to capture the fact that the the msgp size can by 16kb plus the overhead of (up to) 255 arguments. Each additional argument causes the msgp serialization to be a bit bigger in order to encode the length of the argument. I think that means we just remove `config.MaxLogicSIgMaxSize` from the sum being multiplied.

jannotti commented on protocol/tags.go:
```diff
@@ -87,7 +87,7 @@ const TopicMsgRespTagMaxSize = 6 * 1024 * 1024
 // which is size of just a single message containing maximum Stateproof. Since Stateproof
 // transactions can't be batched we don't need to multiply by MaxTxnBatchSize.
-const TxnTagMaxSize = 4620031
+const TxnTagMaxSize = 8460031
```
Hopefully we will not need to change this once we compute the maximum logicsig size correctly.

giuliop replied:
I don't think we can get back to the original value now that we are changing the value of `MaxLogicSigMaxSize` from 1K to 16K.
The current formula is:
```
s = 1 + 2 + msgp.BytesPrefixSize + config.MaxLogicSigMaxSize + 4 + crypto.SignatureMaxSize() + 5 + crypto.MultisigSigMaxSize() + 4
s += msgp.ArrayHeaderSize + ((EvalMaxArgs) * (msgp.BytesPrefixSize + config.MaxLogicSigMaxSize))
```
We are increasing `s` in the first line by 15K, can't really imagine changing the second like in a rational way that reduces it by 15K.
In any case this value will depend on what we decide to do about [this](https://github.com/algorand/go-algorand/pull/6057/#issuecomment-2241226095)

giuliop replied:
Following the implementation of this [comment](https://github.com/algorand/go-algorand/pull/6057#issuecomment-2255765284) we now have:
`const TxnTagMaxSize = 4394756`
which is close but not exactly the original `4620031`.

We could of course define a new constant equal to the original number and use that together with the msgp `maxtotalbytes` directive if we wanted, but I guess there is no reason to use a magic number.

giuliop commented:
> I tentatively agree with you about you thinking about `LogicSigMaxSize()`. There is no explicit constant for a maximum argument size, so an argument can indeed by the whole 16kb in size. But, of course, you can't have 255 of them. So I think we want to capture the fact that the the msgp size can by 16kb plus the overhead of (up to) 255 arguments. Each additional argument causes the msgp serialization to be a bit bigger in order to encode the length of the argument. I think that means we just remove `config.MaxLogicSIgMaxSize` from the sum being multiplied.

I agree with your logic and my first attempt was to do:
```diff
type LogicSig struct {
	_struct struct{} `codec:",omitempty,omitemptyarray"`

	// Logic signed by Sig or Msig, OR hashed to be the Address of an account.
	Logic []byte `codec:"l,allocbound=config.MaxLogicSigMaxSize"`

	Sig  crypto.Signature   `codec:"sig"`
	Msig crypto.MultisigSig `codec:"msig"`

	// Args are not signed, but checked by Logic
-	Args [][]byte `codec:"arg,allocbound=EvalMaxArgs,allocbound=config.MaxLogicSigMaxSize"
+	Args [][]byte `codec:"arg,allocbound=EvalMaxArgs,allocbound=0"`
}
```
This makes msgp generate:
```diff
func LogicSigMaxSize() (s int) {
	s = 1 + 2 + msgp.BytesPrefixSize + config.MaxLogicSigMaxSize + 4 + crypto.SignatureMaxSize() + 5 + crypto.MultisigSigMaxSize() + 4
	// Calculating size of slice: z.Args
-	s += msgp.ArrayHeaderSize + ((EvalMaxArgs) * (msgp.BytesPrefixSize + config.MaxLogicSigMaxSize))
+	s += msgp.ArrayHeaderSize + ((EvalMaxArgs) * (msgp.BytesPrefixSize + 0))
	return
}
```
which implements the idea that the size of the actual arguments is already captured in `config.MaxLogicSigMaxSize` and we only want to add the overhead of the arg len encoding times the max number of args

The issue is that we are saying that the max length on an argument is `0` and this is used also in the `func (z *LogicSig) UnmarshalMsgWithState(...)` which gets broken as shown by the resulting diff:

```diff
diff --git a/data/transactions/msgp_gen.go b/data/transactions/msgp_gen.go
index 7cc22db08..955c1812f 100644
--- a/data/transactions/msgp_gen.go
+++ b/data/transactions/msgp_gen.go
@@ -3306,8 +3306,8 @@ func (z *LogicSig) UnmarshalMsgWithState(bts []byte, st msgp.UnmarshalState) (o
...
-                               if zb0007 > config.MaxLogicSigMaxSize {
-                                       err = msgp.ErrOverflow(uint64(zb0007), uint64(config.MaxLogicSigMaxSize))
+                               if zb0007 > 0 {
+                                       err = msgp.ErrOverflow(uint64(zb0007), uint64(0))
                                        return
                                }
@@ -3395,8 +3395,8 @@ func (z *LogicSig) UnmarshalMsgWithState(bts []byte, st msgp.UnmarshalState) (o
...
-                                       if zb0011 > config.MaxLogicSigMaxSize {
-                                               err = msgp.ErrOverflow(uint64(zb0011), uint64(config.MaxLogicSigMaxSize))
+                                       if zb0011 > 0 {
+                                               err = msgp.ErrOverflow(uint64(zb0011), uint64(0))
                                                return
                                        }
```
My second attempt was to say instead that the max len of the arg array is one max size arg by doing: 
```
Args [][]byte `codec:"arg,allocbound=1,allocbound=config.MaxLogicSigMaxSize"`
```
which generates
`s += msgp.ArrayHeaderSize + ((1) * (msgp.BytesPrefixSize + config.MaxLogicSigMaxSize))`

which also breaks `func (z *LogicSig) UnmarshalMsgWithState(...)` since we are saying the max number of args is 1:
```diff
+++ b/data/transactions/msgp_gen.go
@@ -3287,8 +3287,8 @@ func (z *LogicSig) UnmarshalMsgWithState(bts []byte, st msgp.UnmarshalState) (o
...
-                       if zb0005 > EvalMaxArgs {
-                               err = msgp.ErrOverflow(uint64(zb0005), uint64(EvalMaxArgs))
+                       if zb0005 > 1 {
+                               err = msgp.ErrOverflow(uint64(zb0005), uint64(1))
                                err = msgp.WrapError(err, "struct-from-array", "Args")
                                return
                        }
@@ -3376,8 +3376,8 @@ func (z *LogicSig) UnmarshalMsgWithState(bts []byte, st msgp.UnmarshalState) (o
...
-                               if zb0009 > EvalMaxArgs {
-                                       err = msgp.ErrOverflow(uint64(zb0009), uint64(EvalMaxArgs))
+                               if zb0009 > 1 {
+                                       err = msgp.ErrOverflow(uint64(zb0009), uint64(1))
```

I am out of quick fixes, would you have some other idea? Or shall we just leave it very generous as it is now?  
Alternatively I would need to understand in depth how msgp works and what is go-algorand codec architecture to suggest some other alternatives but it's a non trivial time investment

giuliop committed:
Merge remote-tracking branch 'upstream/master' into pool-max-logicSig-size

giuliop committed:
Implement requested changes

giuliop commented on cmd/goal/clerk.go:
```diff
@@ -986,7 +986,7 @@ func assembleFileImpl(fname string, printWarnings bool) *logic.OpStream {
 		}
 	} else {
-		if uint64(len(ops.Program)) > params.LogicSigMaxSize {
+		if !params.EnableLogicSigSizePooling && uint64(len(ops.Program)) > params.LogicSigMaxSize {
```
I agree, changed.

giuliop committed:
Merge remote-tracking branch 'upstream/master' into pool-max-logicSig-size

giuliop committed:
Implement new max size for LogicSigMaxSize

giuliop commented:
OK, wasn't too bad, studying [algorand/msgp](https://github.com/algorand/msgp) I found an handy `maxtotalbytes` directive which I used to fix the max size of the argument array:
``` diff
@@ -39,7 +39,7 @@ type LogicSig struct {
	// Args are not signed, but checked by Logic
-       Args [][]byte `codec:"arg,allocbound=EvalMaxArgs,allocbound=config.MaxLogicSigMaxSize"`
+       Args [][]byte `codec:"arg,allocbound=EvalMaxArgs,allocbound=config.MaxLogicSigMaxSize,maxtotalbytes=config.MaxLogicSigMaxSize"`
 }
 ```
 I used `config.MaxLogicSigMaxSize` for the max size, which generates:
 ``` diff
 func LogicSigMaxSize() (s int) {
        s = 1 + 2 + msgp.BytesPrefixSize + config.MaxLogicSigMaxSize + 4 + crypto.SignatureMaxSize() + 5 + crypto.MultisigSigMaxSize() + 4
        // Calculating size of slice: z.Args
-       s += msgp.ArrayHeaderSize + ((EvalMaxArgs) * (msgp.BytesPrefixSize + config.MaxLogicSigMaxSize))
+       s += msgp.ArrayHeaderSize + config.MaxLogicSigMaxSize
        return
 }
 ```
 and leaves everything else intact.
 
 Let me know if you like this or would consider something else instead of `config.MaxLogicSigMaxSize`.
 
I like it because that's very close to the max size for the args array (a bit more because the Logic cannot be zero, a bit less because we are not including the overhead from `maxArgs  BytesPrefixSize`).

In any case in the overall `LogicSigMaxSize()` we are double counting `MaxLogicSigMaxSize` for both the Logic and the Args so we have a generous, but reasonable, margin.

jannotti commented:
I was thinking about another approach. How about we just keep the `allocbound` on Args at 1000?  There's no particular reason to need very large args. And, in fact, it's totally pointless for it to be bigger than 4k, since teal values can't be bigger that 4k.  If you tried to access an arg > 4k, you will panic anyway.

giuliop commented:
You made me realize that indeed we need to decide which is the max limit for a logicsig arg.  
Before we had a combined cap for Logic + Args of 1000 which was lower than the AVM limit on teal args of 4096, so it implicitly set the limit at 1000.  
Now that the combined cap is 16000 > 4096 we need to decide what is the limit for args.

My first reaction would be to say 4096 to match the AVM, why artificially lower it after all.

The argument I see in favor of using 1000 is not to change the old value of `TxnTagMaxSize`.  
If there is a concern that we might inadvertently break something then it's more prudent to use 1000 and of course if someone needs to pass a larger arg they can always split it up over multiple args.

In sum, I see an allocbound on arg size of 4096 as the most logical choice, but if maintaining the old `TxnTagMaxSize` is valuable then 1000 is a fine choice too.

jannotti commented:
> You made me realize that indeed we need to decide which is the max limit for a logicsig arg. Before we had a combined cap for Logic + Args of 1000 which was lower than the AVM limit on teal args of 4096, so it implicitly set the limit at 1000. Now that the combined cap is 16000 > 4096 we need to decide what is the limit for args.
> 
> My first reaction would be to say 4096 to match the AVM, why artificially lower it after all.

Yes, we ought to allow 4096 for individual args. I suppose it can be checked in `transactions/data/logic/eval.go` near here:
```
	if cx.txn.Lsig.Args != nil && len(cx.txn.Lsig.Args) > transactions.EvalMaxArgs {
		return false, errTooManyArgs
	}
```
though it's unclear why we don't check this in txn.WellFormed.  It can be an unconditional check (no need to have versioning) because old transactions are implicitly limited to < 1000 bytes.

I dislike that this causes a change in `TxnTagMaxSize` but I don't see a way around it.  It was nice when state proof txn was so much bigger than anything else, we didn't have to adjust `TxnTagMaxSize` for a simple txn addition. In the future, I suppose we will have to do so, so the limit is hit by a "normal" txn, so any increase increases the limit.  The comment explaining `TxnTagMaxSize` will need to be reworked.

There is a unit test that caught this right?

jannotti commented:
I'm really confused by that comment I mentioned.  It says:
```
// TxnTagMaxSize is the maximum size of a TxnTag message. This is equal to SignedTxnMaxSize()
// which is size of just a single message containing maximum Stateproof. Since Stateproof
// transactions can't be batched we don't need to multiply by MaxTxnBatchSize.
```
But that doesn't seem right. `TransactionMaxSize()` seems to bound the size of a single transaction, not stateproof in particular, and it _includes_ the use of all fields, though many are mutually exclusive.

But, it also hints that the TxnTag _should_ take into account 16 "normal" transactions, but doesn't bother to do so, since it's the stateproof txn that really causes the number to blow up.  (I think the TxnTag is used for groups, not single transactions)

jannotti commented:
After further discussion, we're confident that we can make changes to the max size tests so it doesn't fail so easily, without issue.  I'll make those changes in a separate PR.  If you address the rest in the meantime (limit arg length, maybe remove the assembly checks) I'll get the max length changes in.

giuliop committed:
Merge remote-tracking branch 'upstream/master' into pool-max-logicSig-size

giuliop committed:
Make goal more permissive when compiling logicsigs

giuliop committed:
Implement max size for logicsig args of 4096 bytes

giuliop committed:
Make TxnTagMaxSize same as max block size

giuliop commented:
> Yes, we ought to allow 4096 for individual args. I suppose it can be checked in `transactions/data/logic/eval.go` near here:
> 
> ```
> 	if cx.txn.Lsig.Args != nil && len(cx.txn.Lsig.Args) > transactions.EvalMaxArgs {
> 		return false, errTooManyArgs
> 	}
> ```
> 
> though it's unclear why we don't check this in txn.WellFormed.

Done, and I also think that the checks in `func eval` on the number and size of lsig args ought to be in `txn.WellFormed`
``` diff
data/transactions/logic/eval.go
...
+var errLogicSigArgTooLarge = errors.New("LogicSig argument too large")
...
@@ -1305,8 +1306,15 @@ func eval(program []byte, cx *EvalContext) (pass bool, err error) {
        if (cx.EvalParams.Proto == nil) || (cx.EvalParams.Proto.LogicSigVersion == 0) {
                return false, errLogicSigNotSupported
        }
-       if cx.txn.Lsig.Args != nil && len(cx.txn.Lsig.Args) > transactions.EvalMaxArgs {
-               return false, errTooManyArgs
+       if cx.txn.Lsig.Args != nil {
+               if len(cx.txn.Lsig.Args) > transactions.EvalMaxArgs {
+                       return false, errTooManyArgs
+               }
+               for _, arg := range cx.txn.Lsig.Args {
+                       if len(arg) > transactions.MaxLogicSigArgSize {
+                               return false, errLogicSigArgTooLarge
+                       }
+               }

data/transactions/logicsig.go
 // EvalMaxArgs is the maximum number of arguments to an LSig
 const EvalMaxArgs = 255

+// MaxLsigArgSize is the maximum size of an argument to an LSig
+// We use 4096 to match the maximum size of a TEAL value
+// (as defined in `const maxStringSize` in package logic)
+const MaxLogicSigArgSize = 4096
...
@@ -39,7 +44,7 @@ type LogicSig struct {
        Msig crypto.MultisigSig `codec:"msig"`

        // Args are not signed, but checked by Logic
-       Args [][]byte `codec:"arg,allocbound=EvalMaxArgs,allocbound=config.MaxLogicSigMaxSize,maxtotalbytes=config.MaxLogicSigMaxSize"`
+       Args [][]byte `codec:"arg,allocbound=EvalMaxArgs,allocbound=MaxLogicSigArgSize,maxtotalbytes=config.MaxLogicSigMaxSize"`
 }
```

> I dislike that this causes a change in `TxnTagMaxSize` but I don't see a way around it. It was nice when state proof txn was so much bigger than anything else, we didn't have to adjust `TxnTagMaxSize` for a simple txn addition. In the future, I suppose we will have to do so, so the limit is hit by a "normal" txn, so any increase increases the limit. The comment explaining `TxnTagMaxSize` will need to be reworked.
> 
> There is a unit test that caught this right?

Yes, this [one](https://github.com/algorand/go-algorand/blob/8d678c311969979cb0777c3d2befff1fa69621df/node/node_test.go#L783)

giuliop commented on cmd/goal/clerk.go:
```diff
@@ -986,7 +986,7 @@ func assembleFileImpl(fname string, printWarnings bool) *logic.OpStream {
 		}
 	} else {
-		if uint64(len(ops.Program)) > params.LogicSigMaxSize {
+		if len(ops.Program) > config.MaxLogicSigMaxSize {
```
I agree, done

giuliop commented on cmd/goal/clerk.go:
```diff
@@ -1204,6 +1209,10 @@ var dryrunCmd = &cobra.Command{
 				fmt.Fprintf(os.Stdout, "ERROR: %s\n", err.Error())
 			}
 		}
+		lSigMaxPooledSize := len(stxns) * int(params.LogicSigMaxSize)
+		if params.EnableLogicSigSizePooling && lSigPooledSize > lSigMaxPooledSize {
+			reportErrorf("total lsigs size too large: %d > %d", lSigPooledSize, lSigMaxPooledSize)
+		}
```
Agree

giuliop commented:
> After further discussion, we're confident that we can make changes to the max size tests so it doesn't fail so easily, without issue. I'll make those changes in a separate PR. If you address the rest in the meantime (limit arg length, maybe remove the assembly checks) I'll get the max length changes in.

I did implement them before reading your comment, I spent some time thinking about it and my best idea in the end was to use the max block length as the max for the transaction / transaction group message tag (as per below).
I can remove that if you want.

```diff
protocol/tags.go
@@ -84,10 +84,10 @@ const StateProofSigTagMaxSize = 6378
 // Matches  current network.MaxMessageLength
 const TopicMsgRespTagMaxSize = 6 * 1024 * 1024

-// TxnTagMaxSize is the maximum size of a TxnTag message. This is equal to SignedTxnMaxSize()
-// which is size of just a single message containing maximum Stateproof. Since Stateproof
-// transactions can't be batched we don't need to multiply by MaxTxnBatchSize.
-const TxnTagMaxSize = 4394756
+// TxnTagMaxSize is the maximum size of a TxnTag message.
+// Matches current config.MaxTxnBytesPerBlock, the maximum lenght of a block,
+// since a transaction or transaction group cannot be larger than a block.
+const TxnTagMaxSize = 5 * 1024 * 1024

node/node_test.go
@@ -805,7 +804,7 @@ func TestMaxSizesCorrect(t *testing.T) {
        require.Equal(t, ppSize, protocol.ProposalPayloadTag.MaxMessageSize())
        spSize := uint64(stateproof.SigFromAddrMaxSize())
        require.Equal(t, spSize, protocol.StateProofSigTag.MaxMessageSize())
-       txSize := uint64(transactions.SignedTxnMaxSize())
+       txSize := uint64(config.MaxTxnBytesPerBlock)

jannotti commented:
Ok, we merged a looser check that should allow this to pass tests. We came to pretty much the same result (5M limit) by another path.  If you merge and get green checks, I'm ready to approve and I'll talk to others to get another final approval.

giuliop committed:
Revert "Make TxnTagMaxSize same as max block size"

This reverts commit 91d6e09d3d39805d7e5d637e76ab059cae525eff
which is not needed anymore after the merge of:
AVM: Derive looser, but more principled, checks of txn max size #6114

giuliop committed:
Merge remote-tracking branch 'upstream/master' into pool-max-logicSig-size

giuliop committed:
Update TestMaxSizesCorrect with logicsig pooling

giuliop committed:
Don't test app/lsig max size in goal clerk compile

We now let goal clerk compile app program and lsig of any sizes,
even it they will be rejected by the nodes.

giuliop commented:
I like it, i had to make two changes to pass all tests and make all checks green.

The first change was in `TestMaxSizesCorrect`:
``` diff
node/node_test.go
@@ -824,10 +824,10 @@ func TestMaxSizesCorrect(t *testing.T) {
        maxCombinedTxnSize := uint64(transactions.SignedTxnMaxSize())
        // subtract out the two smaller signature sizes (logicsig is biggest, it can *contain* the others)
        maxCombinedTxnSize -= uint64(crypto.SignatureMaxSize() + crypto.MultisigSigMaxSize())
-       // the logicsig size is *also* an overestimate, because it thinks each
-       // logicsig arg can be big, but really the sum of the args and the program
-       // has a max size.
-       maxCombinedTxnSize -= uint64(transactions.EvalMaxArgs * config.MaxLogicSigMaxSize)
+       // the logicsig size is *also* an overestimate, because it thinks that the logicsig and
+       // the logicsig args can both be up to to MaxLogicSigMaxSize, but that's the max for
+       // them combined, so it double counts and we have to subtract one.
+       maxCombinedTxnSize -= uint64(config.MaxLogicSigMaxSize)
```
With lsig pooling the new [max size](https://github.com/algorand/go-algorand/pull/6057/files#diff-8f4a7b812c0892353812d410dcea94ccef1b18277f6c33b4e56437d27c39b134R47) for the lsig arg array is:
[EvalMaxArgs](https://github.com/algorand/go-algorand/pull/6057/files#diff-8f4a7b812c0892353812d410dcea94ccef1b18277f6c33b4e56437d27c39b134R26) (255) * [MaxLogicSigArgSize](https://github.com/algorand/go-algorand/pull/6057/files#diff-8f4a7b812c0892353812d410dcea94ccef1b18277f6c33b4e56437d27c39b134R31) (4096), capped at `config.MaxLogicSigMaxSize` (16 * 1000), that is 16000 which is added to 16000 for the lsig program size so it is double counted and we remove the double counting in `TestMaxSizesCorrect`

The second change is in the integration test [test/e2e-go/cli/goal/expect/tealConsensusTest.exp](https://github.com/algorand/go-algorand/pull/6057/files#diff-9d9a133c6ac373272816e6541cd66e3238317f403f743da0215d489f25341025) where I had to comment out the tests for failing to compile an overly large lsig and an overly large app program since we took out those limitations as per this [comment](https://github.com/algorand/go-algorand/pull/6057#discussion_r1731456444)

jannotti commented on data/transactions/logic/eval.go:
```diff
@@ -1305,8 +1306,15 @@ func eval(program []byte, cx *EvalContext) (pass bool, err error) {
+		if len(cx.txn.Lsig.Args) > transactions.EvalMaxArgs {
+			return false, errTooManyArgs
+		}
+		for _, arg := range cx.txn.Lsig.Args {
+			if len(arg) > transactions.MaxLogicSigArgSize {
```
As a note to self and other reviewers - this test was not present earlier because the max logic sig size of1000, which includes args, meant it wasn't possible for args to be too big.

jannotti commented on data/transactions/logicsig.go:
```diff
@@ -25,6 +25,11 @@ import (
+// MaxLsigArgSize is the maximum size of an argument to an LSig
+// We use 4096 to match the maximum size of a TEAL value
+// (as defined in `const maxStringSize` in package logic)
+const MaxLogicSigArgSize = 4096
```
Should we export that constant and use it directly?

giuliop replied:
It would be nice, but `package logic` already imports `package transactions` and we cannot have circular imports.
We would need to define the constant in `package transactions` and then import it in `package logic`.

Or alternatively make it a consensus parameter; in this case I guess also [const EvalMaxArgs = 255](https://github.com/algorand/go-algorand/pull/6057/files/4de5dd0d3470ac252cc35343438c7ddb21c805ea#diff-8f4a7b812c0892353812d410dcea94ccef1b18277f6c33b4e56437d27c39b134R26) could be made a consensus parameter

jannotti replied:
I'm ok leaving it as is.

jannotti commented on data/transactions/verify/txn.go:
```diff
@@ -225,6 +226,18 @@ func txnGroupBatchPrep(stxs []transactions.SignedTxn, contextHdr *bookkeeping.Bl
+	if groupCtx.consensusParams.EnableLogicSigSizePooling {
+		lSigMaxPooledSize := len(stxs) * int(groupCtx.consensusParams.LogicSigMaxSize)
+		if lSigPooledSize > lSigMaxPooledSize {
+			errorMsg := fmt.Sprintf(
+				"txgroup had %d bytes of LogicSigs, more than the available pool of %d bytes",
+				lSigPooledSize, lSigMaxPooledSize,
+			)
+			err = &TxGroupError{err: errors.New(errorMsg), GroupIndex: -1, Reason: TxGroupErrorReasonNotWellFormed}
```
consider `fmt.Errorf` instead of `fmt.Sprintf` and `errors.New` and you might even not use `err`, so you just get:
```
return nil, &TxGroupError{
     err: fmt.Errorf(...)
     GroupIndex: ...
     Reason: ...
}

giuliop replied:
Yes, that's nicer

jannotti commented on test/e2e-go/cli/goal/expect/tealConsensusTest.exp:
```diff
@@ -54,14 +54,17 @@ if { [catch {
-        eof { ::AlgorandGoal::Abort $expect_out(buffer) }
-        "\n" { ::AlgorandGoal::Abort $expect_out(buffer) }
-    }
+    # we do not test anymore for a max lsig size because we let goal compile any program size,
+    # even if they will be rejected by the nodes
```
This is fine. I had said I wanted to retain the test on old versions, but I was thinking of testing that the network doesn't accept 1001 bytes in old versions.

But I think the tests should be completely deleted, rather than remain as comment fossils for future code paleontologists.

giuliop replied:
I agree, will delete them

jannotti reviewed (commented):
The test you deleted is fine, as I mentioned. But I'm very surprised we don't have an e2e test that confirms the that a too big logicsig is rejected as malformed.

Maybe we _do_ have such a test, but it is being run on a single txn group, so it retains the same behaviour as before.  Could you look around for it, and if found, construct one that exercises a two txn group and the different limits?

giuliop committed:
Add integration test for lsig size pooling

giuliop commented:
> Maybe we _do_ have such a test, but it is being run on a single txn group, so it retains the same behaviour as before. Could you look around for it, and if found, construct one that exercises a two txn group and the different limits?

I did not find that test so I created it [here](https://github.com/algorand/go-algorand/pull/6057/files#diff-8018e89516809642bdc6478498569b4f27c027ae52bec137a61e2e6812e18ac7)
I also had to create this new [network template](https://github.com/algorand/go-algorand/pull/6057/files#diff-456b9eeb77e78c72aea4438860a4aca55619dd7e9e1ce87e532b3b4c7c72e693) to exercise consensus v18 when lsig max size of 1000 was introduced.

jannotti commented on node/node_test.go:
```diff
@@ -824,10 +824,10 @@ func TestMaxSizesCorrect(t *testing.T) {
+	// the logicsig size is *also* an overestimate, because it thinks that the logicsig and
+	// the logicsig args can both be up to to MaxLogicSigMaxSize, but that's the max for
+	// them combined, so it double counts and we have to subtract one.
+	maxCombinedTxnSize -= uint64(config.MaxLogicSigMaxSize)
```
This change is because you tightened the allocbound using the `maxtotalbytes` thing?

giuliop replied:
Yes, now we have:
``` diff
- Args [][]byte `codec:"arg,allocbound=EvalMaxArgs,allocbound=config.MaxLogicSigMaxSize"`
+ Args [][]byte `codec:"arg,allocbound=EvalMaxArgs,allocbound=MaxLogicSigArgSize,maxtotalbytes=config.MaxLogicSigMaxSize"`
```
Without `maxtotalbytes=config.MaxLogicSigMaxSize` it would allocate 255 * 4096 bytes for the args array

jannotti commented on test/e2e-go/features/transactions/logicsig_test.go:
```diff
@@ -0,0 +1,167 @@
+
+// CreateTealOfSize return a TEAL bytecode of `size` bytes which always succeeds.
+// `size` must be at least 9 bytes
+func CreateTealOfSize(size uint, pragma uint) ([]byte, error) {
```
I'd prefer `CreateProgramOfSize()` since it's not the actual TEAL size that matters.

giuliop replied:
Agree, changed

giuliop commented on test/e2e-go/features/transactions/logicsig_test.go:
```diff
@@ -0,0 +1,167 @@
+
+	err = client.BroadcastTransactionGroup([]transactions.SignedTxn{stxn1Fail, stxn2Fail})
+	if cp.EnableLogicSigSizePooling {
+		a.Contains(err.Error(), "more than the available pool")
```
Nice, changed to that

giuliop committed:
Refactor

giuliop commented:
I think this PR is good to go now.

Let me know if you prefer that I move the integration [test](https://github.com/algorand/go-algorand/pull/6057/files#diff-8018e89516809642bdc6478498569b4f27c027ae52bec137a61e2e6812e18ac7) I added to a unit test instead

algorandskiy reviewed (commented):
The change looks reasonable. Please add unit tests to verify/txn_test.go

giuliop committed:
Fix linter warnings

giuliop committed:
Make GenerateProgramOfSize accessible to other packages

giuliop committed:
Add unit test for lsig size

giuliop commented:
> There are two small linter errors that should be fixed. They just appeared because the linter run needed approval to run.

Done, I could have run `make lint` locally to catch them

giuliop commented on data/transactions/verify/txn.go:
```diff
@@ -225,6 +226,17 @@ func txnGroupBatchPrep(stxs []transactions.SignedTxn, contextHdr *bookkeeping.Bl
+				"txgroup had %d bytes of LogicSigs, more than the available pool of %d bytes",
+				lSigPooledSize, lSigMaxPooledSize,
+			)
+			return nil, &TxGroupError{err: errorMsg, GroupIndex: -1, Reason: TxGroupErrorReasonNotWellFormed}
```
I added the unit [test](https://github.com/algorand/go-algorand/pull/6057/files#diff-455b968f2aa356fde47983cabb3239b787eedfc64cf06d807b1df099bc5cf9f7R652), let's see if codecov is happy

giuliop replied:
Yep, it's happy now

giuliop committed:
Add unit test for lsig arg size

giuliop committed:
Fix typo in comments

giuliop commented on data/transactions/verify/txn_test.go:
```diff
@@ -649,6 +649,70 @@ func BenchmarkPaysetGroups(b *testing.B) {
+	consensusVersionPreSizePooling := protocol.ConsensusV18
+	consensusVersionPostSizePooling := protocol.ConsensusFuture
+
+	// We will do tests based on a transaction group of 2 paymentransactions, the first
```
done

giuliop committed:
Merge remote-tracking branch 'upstream/master' into pool-max-logicSig-size

giuliop commented:
All looks good to me now; codecov shows some lines not covered in `cmd/goal/clerk.go` but I suppose that's ok; let me know otherwise

giuliop commented:
I've seen that this PR fails the type serialization checking with `go-algorand-sdk` so I made this [PR](https://github.com/algorand/go-algorand-sdk/pull/659) to `go-algorand-sdk` to hopefully fix it.

Below the failing case:

```
Compare the Type Trees "config.ConsensusParams" v "config.ConsensusParams"

========================================================
			STRUCT DIFF REPORT
comparing
	<<<<<github.com/algorand/go-algorand/config :: "config.ConsensusParams" (struct)>>>>>
VS
	<<<<<github.com/algorand/go-algorand-sdk/v2/protocol/config :: "config.ConsensusParams" (struct)>>>>>
========================================================
--------------------------------------------------------
		DIFFERENCES FOUND (partial)
--------------------------------------------------------
Common path of length 0:

X-DIFF
------
EXISTS IN:	 	"github.com/algorand/go-algorand/config :: \"config.ConsensusParams\" (struct)"
MISSING FROM:		"github.com/algorand/go-algorand-sdk/v2/protocol/config :: \"config.ConsensusParams\" (struct)"
1 TYPES TOTAL:

(1)
[FIELD](+codec): 	[EnableLogicSigSizePooling]()
SOURCE: 		 :: "bool" (bool)
```
